embassy-usb-driver/src/host.rs

```diff
@@ -0,0 +1,309 @@
+        }
+        fn is_out() -> bool {
+            false
+        }
+    }
+    impl Direction for Out {
+        fn is_in() -> bool {
+            false
+        }
+        fn is_out() -> bool {
+            true
+        }
+    }
+    impl Direction for InOut {
+        fn is_in() -> bool {
+            true
+        }
+        fn is_out() -> bool {
+            true
+        }
+    }
+
+    #[diagnostic::on_unimplemented(message = "This is not an IN channel")]
+    pub trait IsIn: Direction {}
+    impl IsIn for In {}
+    impl IsIn for InOut {}
+
+    #[diagnostic::on_unimplemented(message = "This is not an OUT channel")]
+    pub trait IsOut: Direction {}
+    impl IsOut for Out {}
+    impl IsOut for InOut {}
+}
+
+/// Specify the timeout of a channel
+pub struct TimeoutConfig {
+    /// Maximum response timeout for transactions with a Data Stage
+    pub data_timeout: Duration,
+    /// Maximum response timeout for transactions without a data stage
+    pub standard_timeout: Duration,
+}
+
+impl Default for TimeoutConfig {
+    fn default() -> Self {
+        TimeoutConfig {
+            data_timeout: Duration::from_millis(500),
+            standard_timeout: Duration::from_millis(50),
+        }
+    }
+}
+
+/// ## Virtual USB Channels
+/// These contain the required information to send a packet correctly to a device endpoint.
+/// The information is carried with the channel on creation (see [`UsbHostDriver::alloc_channel`]) and can be changed with [`UsbChannel::retarget_channel`].
+///
+/// It is up to the hal's driver how to implement concurrent requests, some hardware IP may allow for multiple hardware channels
+///  while others may only have a single channel which needs to be multiplexed in software, while others still use DMA request linked-lists.
+/// Any of these are compatibile with the UsbChannel with varying degrees of sync primitives required.
+pub trait UsbChannel<T: channel::Type, D: channel::Direction> {
+    /// Send IN control request
+    async fn control_in(&mut self, setup: &SetupPacket, buf: &mut [u8]) -> Result<usize, ChannelError>
+    where
+        T: channel::IsControl,
+        D: channel::IsIn;
+
+    /// Send OUT control request
+    async fn control_out(&mut self, setup: &SetupPacket, buf: &[u8]) -> Result<usize, ChannelError>
+    where
+        T: channel::IsControl,
+        D: channel::IsOut;
+
+    /// Retargets channel to a new endpoint, may error if the underlying driver runs out of resources
+    fn retarget_channel(&mut self, addr: u8, endpoint: &EndpointInfo, pre: bool) -> Result<(), HostError>;
+
+    /// Send IN request of type other from control
+    /// For interrupt channels this will return the result of the next succesful interrupt poll
+    async fn request_in(&mut self, buf: &mut [u8]) -> Result<usize, ChannelError>
+    where
+        D: channel::IsIn;
+
+    /// Send OUT request of type other from control
+    async fn request_out(&mut self, buf: &[u8]) -> Result<usize, ChannelError>
```

Comment:
Why does this return `usize`? Would the channel ever write less data that what was given and not call it some kind of error?

Reply:
True, had to check the other drivers (I'm most familiar with synposys), but they all do full writes, so the usize is unnecessary.  I also don't imagine a driver would fall back to a higher layer rather than looping & yielding. So I'll change this.